### PR TITLE
Use converters when recording a bag file

### PR DESCRIPTION
**Pull request by botteroa-si**
This PR is based on #56. It includes:
- the addition of the option `-f, --serialization-format` for the CLI verb `record`
- the `serialization-format` option defaults to the `rmw` format of the incoming messages
- if a conversion is needed, the appropriate converter plugin is used during the record process, and if such plugin does not exists, the record process fails gracefully
- now each topic has its own `rmw` serialization format (for the moment, though, a bag file can be played only if all topics have the same format. Since it is not yet possible to edit a bag file, this is not a problem. The possibility to play a bag with different serialization formats will be added in a followup PR)

### CODE REVIEW

**Comment by anhosi** [comment]
Fixes #16

**Comment by anhosi** [comment]
CI
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5645/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2262/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4670/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5541/)

**Comment by Martin-Idel-SI** [comment]
This PR is now based on #64 - after this PR is merged, the number of files changed will shrink to 63.

The changeset still comprises a lot of files since we change the behaviour of the storage plugin: Instead of writing one topic type per storage storage, now every topic has its own type. This needs to be changed in the sqlite3 backend all the way to the printing statements.

File: rosbag2/include/rosbag2/writer.hpp [hunk 1]
```diff
@@ -45,17 +48,26 @@ class ROSBAG2_PUBLIC Writer
...
   explicit
   Writer(
-    std::unique_ptr<rosbag2_storage::StorageFactoryInterface> factory =
-    std::make_unique<rosbag2_storage::StorageFactory>());
+    std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory =
+    std::make_unique<rosbag2_storage::StorageFactory>(),
+    std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory =
+    std::make_shared<SerializationFormatConverterFactory>(),
+    std::unique_ptr<rosbag2_storage::MetadataIo> metadata_io =
+    std::make_unique<rosbag2_storage::MetadataIo>());
```
Comment (Karsten1987) [inline rosbag2/include/rosbag2/writer.hpp]:
```suggestion
    std::make_unique<rosbag2_storage::MetadataIo>());
    
```

File: rosbag2/src/rosbag2/converter.cpp [hunk 2]
```diff
@@ -28,16 +28,18 @@
 namespace rosbag2
 {
 
+struct ConverterTypeSupport
```
Comment (Karsten1987) [inline rosbag2/src/rosbag2/converter.cpp]: I believe this deserves some docs here. What's the difference between `type_support` and `introspection_type_support`? Does it make sense to rename the `type_support` into something more specific?

File: rosbag2/include/rosbag2/converter.hpp [hunk 3]
```diff
@@ -60,11 +62,13 @@ class ROSBAG2_PUBLIC Converter
...
   convert(std::shared_ptr<const SerializedBagMessage> message);
 
+  void add_topic(const std::string & topic, const std::string & type);
+
 private:
   std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory_;
   std::unique_ptr<SerializationFormatConverterInterface> input_converter_;
   std::unique_ptr<SerializationFormatConverterInterface> output_converter_;
-  std::map<std::string, std::string> topics_and_types_;
+  std::map<std::string, ConverterTypeSupport> topics_and_types_;
```
Comment (Karsten1987) [inline rosbag2/include/rosbag2/converter.hpp]: does this have to be an ordered map? What about `std::unordered_map`?

File: rosbag2/src/rosbag2/converter.cpp [hunk 4]
```diff
@@ -58,9 +60,8 @@ Converter::~Converter()
 std::shared_ptr<SerializedBagMessage> Converter::convert(
   std::shared_ptr<const rosbag2::SerializedBagMessage> message)
 {
-  auto ts = get_typesupport(topics_and_types_[message->topic_name], "rosidl_typesupport_cpp");
-  auto introspection_ts =
-    get_typesupport(topics_and_types_[message->topic_name], "rosidl_typesupport_introspection_cpp");
+  auto ts = topics_and_types_.at(message->topic_name).type_support;
```
Comment (Karsten1987) [inline rosbag2/src/rosbag2/converter.cpp]: `at` might throw. Are we positive that the map is always filled at this point?

File: rosbag2/src/rosbag2/converter.cpp [hunk 5]
```diff
@@ -72,4 +73,14 @@ std::shared_ptr<SerializedBagMessage> Converter::convert(
   return output_message;
 }
 
+void Converter::add_topic(const std::string & topic, const std::string & type)
+{
+  ConverterTypeSupport type_support;
+  type_support.type_support = get_typesupport(type, "rosidl_typesupport_cpp");
```
Comment (Karsten1987) [inline rosbag2/src/rosbag2/converter.cpp]: I think I'd prefer having this renamed into something like `rmw_type_support` or something.

File: rosbag2/src/rosbag2/sequential_reader.cpp [hunk 6]
```diff
@@ -41,13 +43,29 @@ SequentialReader::open(const StorageOptions & options, const std::string & rmw_s
   if (!storage_) {
     throw std::runtime_error("No storage could be initialized. Abort");
   }
-  auto storage_serialization_format = storage_->get_metadata().serialization_format;
+  auto topics = storage_->get_metadata().topics_with_message_count;
+  if (topics.empty()) {
+    return;
+  }
+
+  // Currently a bag file can only be played if all topics have the same serialization format.
```
Comment (Karsten1987) [inline rosbag2/src/rosbag2/sequential_reader.cpp]: I thought we just introduced a mapping from topic to serialization format?

File: rosbag2/CMakeLists.txt [hunk 7]
```diff
@@ -119,11 +121,36 @@ if(BUILD_TESTING)
...
-    test/rosbag2/types/test_ros2_message.cpp)
+    test/rosbag2/types/test_ros2_message.cpp
+    src/rosbag2/typesupport_helpers.cpp
+    src/rosbag2/types/ros2_message.cpp)
   if(TARGET test_ros2_message)
-    target_link_libraries(test_ros2_message rosbag2)
-    ament_target_dependencies(test_ros2_message test_msgs)
+    target_compile_definitions(test_ros2_message PRIVATE "ROSBAG2_BUILDING_DLL")
+    if(NOT DISABLE_SANITIZERS)
+      target_compile_options(test_ros2_message PUBLIC $<$<CXX_COMPILER_ID:GNU>:-fsanitize=leak>)
```
[the remark this answers is not in the document]
Comment (anhosi) [reply rosbag2/CMakeLists.txt]: It did not work on the Mac of @botteroa-si.

File: rosbag2/src/rosbag2/converter.cpp (hunk 2 above)
Comment (anhosi) [reply rosbag2/src/rosbag2/converter.cpp]: I like your suggestion below `rmw_type_support`.

File: rosbag2/src/rosbag2/converter.cpp (hunk 4 above)
Comment (anhosi) [reply rosbag2/src/rosbag2/converter.cpp]: Yes we are. `add_topic` must be (and is) called for every topic before a message of such a topic can be converted. Failing to do so would be a implementation error so a exception would be ok.

File: rosbag2/include/rosbag2/converter.hpp (hunk 3 above)
Comment (anhosi) [reply rosbag2/include/rosbag2/converter.hpp]: Correct a `unordered_map` is sufficient. This is true for almost all our usages of `std::map`. As this refactoring will touch a lot of files I did it as separate pr #65.

File: rosbag2/src/rosbag2/converter.cpp (hunk 2 above)
Comment (Karsten1987) [reply rosbag2/src/rosbag2/converter.cpp]: but it will fetch the type support from the middleware, doesn't it?

Comment (anhosi) [reply rosbag2/src/rosbag2/converter.cpp]: As discussed we will rename it in a follow-up pr-

File: rosbag2_transport/include/rosbag2_transport/record_options.hpp [hunk 8]
```diff
@@ -25,6 +25,7 @@ struct RecordOptions
 public:
   bool all;
   std::vector<std::string> topics;
+  std::string rmw_serialization_format;
```
[the remark this answers is not in the document]
Comment (anhosi) [reply rosbag2_transport/include/rosbag2_transport/record_options.hpp]: I would do this only when this feature is being implemented. I am not yet sure how this should look like (or even how the cli usage should be). E.g. a map might also be an alternative.

File: rosbag2/CMakeLists.txt [hunk 9]
```diff
@@ -119,11 +121,36 @@ if(BUILD_TESTING)
...
+      target_link_libraries(test_ros2_message $<$<CXX_COMPILER_ID:GNU>:-fsanitize=leak>)
```
Comment (dirk-thomas) [inline rosbag2/CMakeLists.txt:133]: Just for the record: in the Jenkins CI jobs which are using Docker the leak sanitizer was never used for the test. See ros-infrastructure/ros_buildfarm#832.

**Pull request merged**